PEP 773: A Python Installation Manager for Windows (#4219)
python/peps: peps/pep-0772.rst

PEP: 772
Title: A Python Installation Manager for Windows

> Author: Steve Dower
> PEP-Delegate: TBD
> Discussions-To: https://discuss.python.org/t/pep-772-a-python-installation-manager-for-windows/77900
> Status: Draft
> Type: Standards Track
> Topic: Release
> Created: 21-Jan-2025
> Post-History: 18-Dec-2024

Reviewer: We can link to Post-History now, but I don't have the link for your old topic -- please edit the suggestion to add it!
```suggestion
Author: Steve Dower
Discussions-To: https://discuss.python.org/t/77900/
Status: Draft
Type: Standards Track
Topic: Release
Created: 21-Jan-2025
Post-History:
   `18-Dec-2024 <https://discuss.python.org/t/77900/>`__
   `21-Jan-2025 <https://discuss.python.org/t/77900/>`__
```

Abstract
========

Installation of the python.org Python distribution on Windows is complex.
There are three main approaches with roughly equivalent levels of user
experience, and yet all of these suffer from different limitations, including
failing to satisfy modern usage scenarios. This PEP proposes a design for
a single Windows install workflow tool that satisfies all the needs of the
existing installers for the platform, while avoiding most of their limitations,
and provides the core team with the ability to manage releases for many years
to come.

Reader's Note
-------------

This document is a detailed *proposal*, intended to assist in an "approve" or
"reject" decision on the replacement of the existing install formats for Python
on Windows with a new format. It is not intended to be a binding specification,

> nor is it user documentation or an interoperability specification. The actual

Reviewer: As such, should this be `Type: Informational` rather than `Type: Standards Track`?

Author: Open for discussion, but generally "Informational" PEPs never reach "Final", and this one should. Plus it's somewhat tied to a release, even though it's not tied to the language.

implementation may vary over time as needs change, and that is not to be
considered a "violation" of this PEP. Any interfaces and protocols meant for
public use will be independently documented and maintained in accordance with
standard deprecation timelines.

Documentation on installing Python on Windows can be found at

> `docs.python.org/using/windows.html <https://docs.python.org/using/windows.html>`__.

Reviewer: We can use Sphinx roles here. We could also change the link text to something more meaningful than just the URL?
```suggestion
:ref:`docs.python.org/using/windows.html <using-on-windows>`.
```

Author: The point is to show the exact URL, so that people feel confident linking it from other places. Otherwise I've seen some people get nervous that they won't be able to point people towards documentation.

* most users do not want to use administrator privileges
* most users will benefit from installing maintenance updates
* most users expect the ``python`` command to work after installation

The primary support for these assertions is that the most popular installers
actively chosen by users are the latest stable release on python.org, and
the latest stable release on the Windows Store, both of which meet these
requirements.

We make the following assumptions about other significant sets of users.
These may have some overlap between groups, and at least some users expect
all of them.

* some users want to install Python programmatically
* some users want to install a particular version
* some users want to install many versions
* some users want to install for all users of their machine
* some users do not want Start Menu shortcuts
* some users want to install as part of their project's build process
* some users want to install as part of their project's install process
* some users intend to never update their install
* some users expect the ``py`` command to work after installation

These assumptions have all been demonstrated over time to exist, though the

> relative importance has not been quantified. The Nuget packages and the
> embeddable distro can meet most of these needs.

Reviewer: Could we link to these two?

```suggestion
relative importance has not been quantified. The `Nuget packages`_ and the
:ref:embeddable distro <windows-embeddable>` can meet most of these needs.

.. _Nuget packages: https://www.nuget.org/packages/python
```

Author: ```suggestion
relative importance has not been quantified. The :ref:`NuGet packages
<windows-nuget>` and the :ref:`embeddable distro
<windows-embeddable>` can meet most of these needs.
```

Traditional Installer
---------------------

The traditional installer is an executable downloadable directly from
python.org that installs the entire development kit for Python. This includes
the CPython interpreter, the standard library, Python headers and import
libraries, builds of Tcl and Tk, the documentation as HTML files, the runtime
and standard library test suite, Start Menu shortcuts for Python and IDLE,
debugging symbols and debug builds of the binaries, the ``py.exe`` launcher
and its file associations, and functionality to modify the user's ``PATH``
environment variable, enable long-path support on their system, pre-generate

> ``.pyc`` files for the standard library, and install pip.

Reviewer: ```suggestion
``.pyc`` files for the standard library, and a ``pip`` installation.
```

Author: pip doesn't quote their own name, so I chose not to.

provides the most direct method of using Windows Installer. The bundle is a
custom C++ application, based on their template, which allows us to customise
the overall behaviour of the installer to determine precisely which MSI files
should actually be installed. The process of copying files, updating the
registry, and generating shortcuts is handled entirely by Windows Installer.

As well as the intended uses, it is understood that many users will (attempt to)
use the traditional installer for other scenarios, such as unregistered installs
and automated CI system installs. While better alternatives are available, they
are not as obvious, and the hope is that a future design would make these
scenarios easier.

Windows Store
-------------

The Windows Store packages for CPython are produced as part of our normal
release process using almost all identical binaries to the other packages.
Due to being in an app store package, the primary ``python.exe`` is enhanced
to be able to determine its location properly, and alternative ``pip.exe`` and
other shortcuts are included to make up for the lack of ``PATH`` environment

> variable settings. These are implemented in ``PC\\python_uwp.cpp`` in our repo.

Reviewer: ```suggestion
variable settings. These are implemented in :file:`PC\\python_uwp.cpp` in our repo.
```

These packages are installed by searching for Python in the Microsoft Store
app, which will find results for each major version since 3.8. Users then have
to select a version and install it. These packages include the CPython

> interpreter, standard library, Tcl/Tk, IDLE and pip, and create file

Reviewer: ```suggestion
interpreter, standard library, Tcl/Tk, IDLE, and pip, and create file
```

such as virus scanners, other installers, system configuration, admin policies,
and even other files in the same directory as the installer. On top of this,
most of its advanced and beneficial functionality such as update patches,
incremental updates, and automatic rollback are unimportant for Python users.

Most user expectations are *defined* by the traditional installer, and so by
definition, it meets them. One primary gap is that it is not able to create an
"unmanaged" install - that is, the equivalent of only copying files onto the
user's system without registration. If you have installed it once, and you
try to install it again, you will only even be able to manage (or upgrade) the
existing install. This can lead to installs moving on update, which will
break users.

Additionally, the ``PATH`` environment variable cannot be intelligently
modified - at best, we can prepend or append the install path. This usually
results in the most recent install of Python being the highest priority. For
example, if the user has Python 3.14 installed and then installs (or updates)
3.13, the ``python`` command will switch from the later version to the earlier
version.

> The ``py.exe`` launcher, defined in PEP 397 and implicitly updated by PEP 514,

Reviewer: ```suggestion
The ``py.exe`` launcher, defined in :pep:`397` and implicitly updated by :pep:`514`,
```

version expected by the user.

The Windows Store package is very reliable, with the exception of the global
shortcuts. Rather than modifying ``PATH`` to add its own directory, these
shortcuts are created in a single OS managed directory that has all the
shortcuts defined by any app. Users are able to modify their ``PATH`` to exclude
or de-prioritise this directory, leading to unreliable or inconsistent
behaviour, and historically we have also seen this caused by installers.
For example, installing Python from the Store followed by Python from the
traditional installer with its ``PATH`` modification enabled will almost always
shadow the Store package's Python with the later install.

User expectations that are un-met by the Store package tend to be performance
and technical. Due to the overhead of launching an app, Python starts up slower.
Because apps are designed to be isolated from each other, it is more difficult
to use hidden directories (such as ``AppData`` or ``TEMP``) to communicate
between different versions of Python, as each version has its own space. Apps
are subject to stricter security requirements that legacy applications start

> disabled, such as DLL hijacking protection, which causes some libraries to fail.

Reviewer: "that legacy applications start disabled" reads oddly, I'm not sure what is intended here.

Python Software Foundation. After publishing, we will request that Microsoft
adjust their ``python.exe`` stub to open to this new app.

This app does not directly provide a version of Python, but it does provide the
global commands that users expect to work, as well as file associations and
Start menu shortcuts. The OS will prompt users to launch the app after install,
which will trigger an automatic install of the current release of CPython and
then launch it. From the user's perspective, they have the same initial
experience as today, with one added progress bar on first launch.

The global commands provided by the app must be static and bundled into the app
itself. They can only change their behaviour at runtime, and cannot be
redirected to different executables except by the user (and then only to another
installed app). So the commands to be provided by PyManager are ``python.exe``,
``python3.exe``, ``py.exe``, ``pymanager.exe``. Each of these must have the
ability to inspect the user's system and choose the correct runtime to launch.
Additionally, ``py`` and ``pymanager`` will have management subcommands to allow
adding and removing runtimes.

> In line with PEP 394 and the default behaviour of Windows, the recommended
> command for launching Python is ``python.exe``. As provided by PyManager, this
> will locate an existing install, either among those that PyManager manages or
> using PEP 514, or it will install the latest available version of CPython and
> select that. The ``python3.exe`` command behaves similarly, but is only allowed
> to find 3.x installs from python.org.

Reviewer: ```suggestion
In line with :pep:`394` and the default behaviour of Windows, the recommended
command for launching Python is ``python.exe``. As provided by PyManager, this
will locate an existing install, either among those that PyManager manages or
using :pep:`514`, or it will install the latest available version of CPython and
select that. The ``python3.exe`` command behaves similarly, but is only allowed
to find 3.x installs from python.org.
```

The ``py.exe`` command provided by PyManager will be recommended for most
management use, due to its brevity. ``py install ...``, ``py list ...`` and so
on. The proposed commands are detailed later. The existing behaviour of the

> PEP 397 launcher is preserved, however, launching through ``py`` will not

Reviewer: ```suggestion
:pep:`397` launcher is preserved, however, launching through ``py`` will not
```

precedence over these commands, and so these are a last resort in place of an
error message. As a result, we can generally assume that a user is launching
these commands because they haven't configured a stronger preference (for
example, a user who has activated an environment will never launch our
``python.exe``, because activation will put a different one ahead of it, and a
user who wants precisely the behaviour of the existing ``py.exe`` can just
install it and will never launch our new one).

The ``pymanager.exe`` command is to allow for handling ambiguous situations.
Existing installs of Python and the launcher may shadow ``python.exe`` and
``py.exe``, but in an automated environment, this can make administrative
scripts unreliable, and so the ``pymanager`` command is unlikely to refer to
something other than PyManager. It has all the subcommands, and launching it
with no command specified will print help for the user.

Replacing other installers
--------------------------

Our intent is to immediately stop publishing individual versions to the Windows
Store, and to deprecate and phase out the traditional installer and Nuget

> packages by 3.16. The embeddable distro would remain, but its listing on

Reviewer: ```suggestion
packages by Python 3.16. The embeddable distro would remain, but its listing on
```

   py install [-s|--source <URL>] [-d|--download <DIR>] [tag ...]
   py install --refresh

.. note::
   This and all later subcommands are also available under ``pymanager``.
   However, as we intend for ``py`` to be the usual command, we only show that
   one.

This subcommand will install one or more runtimes onto the current machine.
The tags are ``Company\\Tag`` pairs (or just ``Tag`` if no slash is included),
and are used to search the index file. Company names match as case-insensitive
prefixes, preferring a full match over a prefix, and tags use case-insensitive,
number-aware matches, with dotted numbers treated as versions. Tags must match
one of the listed "install for" tags, and entries list multiple such tags to
handle abbreviated requests.

For example, the ``3.10.5`` entry would list all of ``3``, ``3.10`` and
``3.10.5`` as tags to be installed for. A request for ``3.10`` would match one
of these and so the entry is selected. Due to the number-aware matches, a
request for ``03.0010`` would also match, and ``3.10.50`` would not.

Tags may also be specified as a constraint, using ``>``, ``>=``, ``<``, ``<=``

> or ``!=`` followed by the ``Company\\Tag`` or ``Tag`` value. When matching a

Reviewer: Is `~=` allowed?

Author: Great question (for the discussion thread ;) ) (but not yet, and not planned, unless someone suggests/motivates it)

any Python runtime.

The index is stored as JSON. The main top level key is ``versions``, which
contains a list of objects. Each version object has its own schema version, and
there is no overall file schema version. Future changes may add additional
top-level keys to provide functionality that cannot be safely integrated into
an existing one.

Version objects may be split between the index file and a ``__install__.json``
stored in the root of each package archive. The entries in the bundled file will
fill in any gaps from the index file. This is intended to allow the typically
large ``shortcuts`` key to be removed from the index file, but may also extend
to ``alias``, ``executable`` and ``executable_args``. Omitting other keys from
the index may result in problems installing the package.

A second top-level key ``next`` contains an optional URL to another index. This
may be used if PyManager cannot find a suitable package in the included
versions. The intent is to allow for older indexes to be archived and only
accessed when required, reducing the size of the initial download without
cutting off users from older versions.

> The initial schema is shown below::
> 
>    SCHEMA = {

Reviewer: ```suggestion
The initial schema is shown below:

.. code:: python

   SCHEMA = {
```

not allow for prerelease or experimental builds to be treated specially, and so
``py.exe`` often prefers these builds by default over the non-experimental
version expected by the user.

The Windows Store package is very reliable, with the exception of the global
shortcuts. Rather than modifying ``PATH`` to add its own directory, these
shortcuts are created in a single OS managed directory that has all the
shortcuts defined by any app. Users are able to modify their ``PATH`` to exclude
or de-prioritise this directory, leading to unreliable or inconsistent
behaviour, and historically we have also seen this caused by installers.
For example, installing Python from the Store followed by Python from the
traditional installer with its ``PATH`` modification enabled will almost always
shadow the Store package's Python with the later install.

User expectations that are un-met by the Store package tend to be performance
and technical. Due to the overhead of launching an app, Python starts up slower.
Because apps are designed to be isolated from each other, it is more difficult
to use hidden directories (such as ``AppData`` or ``TEMP``) to communicate

> between different versions of Python, as each version has its own space. Apps
> are subject to stricter security requirements that legacy applications start
> disabled, such as DLL hijacking protection, which causes some libraries to fail.

Author: ```suggestion
between different versions of Python, as each version has its own space. Apps
are subject to stricter security requirements that legacy applications usually
have disabled, such as DLL hijacking protection, which causes some libraries to fail.
```

become the preferred method for most users, with many preferring the global
modifications to the ``PATH`` environment variable. However, the command itself
has come to be relied upon, and should be preserved as long as possible. This is
achieved in two ways.

Firstly, we install our own ``py.exe`` alias with PyManager that provides the
same functionality, along with PyManager specific functionality. This is
intended to become the default/preferred install of ``py.exe`` over time.

Second, we generate PEP 514 metadata (when requested) for each install, which
allows a legacy ``py.exe`` to continue to work normally with installs managed by
PyManager.

Due to how the existing ``py.exe`` launcher configures itself, and how the MSIX
package for PyManager is constrained, it is not possible for PyManager's ``py``
alias to override the launcher. As a result, users who install the launcher will
always find ``py`` resolving to the launcher. Ultimately, the only way to
resolve this in favour of PyManager is to uninstall the launcher, which can be
done through the standard Installed Apps control panel.

Interaction with venv

> ---------------------

Author: Specifically for `venv`, not anyone else's implementation of it. Those are outside of my scope here.

Users who activate a corrupt or misconfigured virtual environment that is either
missing its ``python.exe`` or has it not on ``PATH`` may receive a different
error from before.

PyManager's global ``python`` alias will be found and executed instead,
suppressing any system "not found" error. As it fails to find the environment's
actual runtime, it will then fail, though the code and message may be different.

As this scenario requires an already corrupt system, this change will be
documented only.

Old version availability
------------------------

Python versions prior to the first release of PyManager can be backfilled into
the python.org index, either based on newly repackaged archives or using the
almost equivalent packages from Nuget (the latter does not include Tcl/Tk,
making them significantly incompatible for some users, but this is likely okay
for especially old versions).

Administrator installs
----------------------

Installing a copy of Python for all users is no longer possible, as PyManager
will only install into the user's own directory. No scenario has been presented
to show that per-machine installs are in line with our intent for the upstream

> distribution, and so we will simply not provide an option for them. Third

Author: Luckily, `py` is shorter than both - and you can even `py -3.11` if you need to be specific!

> distribution, and so we will simply not provide an option for them. Third

Reviewer: To be fair, the only time I need to write the path is when using `uv venv --python=...` -- I do normally use `py -3`!

> distribution, and so we will simply not provide an option for them. Third

Author: A new trick here (assuming Powershell) would be `uv venv "--python=$(py list -1 -f=exe 3.11)"`

Or I expect uv will automatically install our builds anyway, so might as well just let it :)

sufficient instruction for the most basic cases. Thanks to the redirector added
by Microsoft, following this instruction will at least result in something
useful happening, and with PyManager we can ensure that "something useful" means
that the user is running the latest version.

To explain what is actually happening, we propose the following as introductory
text:

.. code:: text

   Python installs on Windows are managed using an installer tool. After it has
   been installed, you can run ``python`` to launch the interpreter, and it will
   choose the best version already installed, available online, or referenced by
   the script you are launching (if any). If you have a preference for a
   particular version, you can specify it with ``py -V:<version>`` followed
   by the rest of your command.

   To install a version of Python without running any command, use ``py install
   <version>``. You can see all of your installs with ``py list`` and remove them
   with ``py uninstall <version>``. Run ``py help`` to see all the options that
   are available.

   Because each version of Python will be shared by all your projects, we

>    recommend using virtual environments. This will usually be created for a

Author: This is only a suggestion, to help save people from having to imagine what they may write on their own. I don't really expect people to copy-paste this.

The text below for projects really does require adding the link, because nobody wants to explain how to install Python on every OS, and it seems many aren't aware that we've already done it.

we sign the python.org package ourselves, while the Store package is signed as
part of the publish process. Otherwise, there is no additional cost to producing
and publishing both packages.

Inline Script Metadata
----------------------

PEP 723 introduced inline script metadata, a structured comment intended for
third-party tools to interpret and then launch a Python script in the correct
environment. An example taken from that PEP:

   # /// script
   # requires-python = ">=3.11"
   # dependencies = [
   #   "requests<3",
   #   "rich",
   # ]
   # ///

PyManager has no integrated support for installing dependencies, and does not
propose adding any. As a result, we could not fully implement handling of this
metadata, and as we consider partial handling to be worse than nothing, we
choose not to implement any.

It is possible for a user to specify the constraint directly as an option, for
example, ``py -V:>=3.11 my-script.py`` to get the selection behaviour.

We could also detect the metadata and warn if the selected runtime does not
match its requirement, but this is not part of the initial proposal.

> 

Author: I'll commit it, but I'm not sure if this change is going to increase or decrease the number of blank lines...

> PEP: 772

Reviewer 2: Please use 773, https://github.com/python/peps/pull/4218 pipped you to 772 by a couple of hours. 
```suggestion
PEP: 773
```

``python`` command, then the redirection will remain in place. In order to
ensure that users do not get stuck always going back to the Store, we need to
provide this command. (The same applies to ``python3``.)

Second, the alternative to the "not real" alias is not "the real" one. It's
nothing. We don't have the ability to replace the global static alias with one
that follows the user's preference or installs, and so the alternative would be
to provide nothing and have ``python`` be an error in all cases. This is worse,
and in our opinion, actively harmful to Python's reputation.

Third, although the underlying implementation of the ``python`` alias is more
complex than the default ``Programs/python.c``, the experience of using it is
identical. The alias is only launched in the absence of another expressed
preference (that is, there's nothing else on ``PATH``), it respects any
indirect preferences (such as through configuration or shebangs), and it
launches the most appropriate version of Python available on the user's machine.
This is much closer to the desired behaviour of the global ``python`` command

> than any alternative.

Reviewer 3: I like this proposed behavior a lot. In fact, it's similar to the behavior of the global `python` command on Gentoo Linux. Maybe it's worth calling that out as an additional comparison here?

For context, on Gentoo Linux, the following executables are symlinks to [`python-exec2c`](https://github.com/projg2/python-exec/):
- `/usr/bin/python`
- `/usr/bin/python3`

There is a configuration file (`/etc/python-exec/python-exec.conf`) as well as an environment variable (`$EPYTHON`) which controls "preferred order" of python installations. When run directly (as `python` or `python3`) or when used as a shebang, the program will check for the real program in `/usr/lib/python-exec/{IMPLEMENTATION}/` and run it, using your preferred implementation or falling back to an alternative implementation (so for example if you have set the environment variable to "python3.12" but only installed a script for python 3.13, it will prefer to run the script and not return an error). Using an invalid value might e.g. print:

```
$ EPYTHON=python3.15 python
python-exec: EPYTHON value invalid (python3.15).
Python 3.12.8 (main, Jan 19 2025, 12:42:41) [GCC 14.2.1 20241221] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> 
```

This is a slightly meandering comment, hmm. Anyways, my point is that this isn't simply a good idea to better serve Windows users. It's a good idea that has prior art. Nobody has ever complained that we (Gentoo) replaced their python program with a "fake". In fact, they haven't even complained that the interpreter sometimes prints a non-CPython warning message before starting up. The executable does, as you say, provide the identical *experience* of a python.

"python" is defined to "run some version of python" and it does that however it pleases, using whatever internal means it chooses to decide which python you meant. If people wanted a stronger preference, they would simply run it themselves. It seems to be working out so far. We should spread this convention to other systems. :tada:

Reviewer 3: Well, for what it's worth Gentoo's python executable doesn't have install capabilities (you still have to run the real package manager for that, though I suppose there's no reason pymanager couldn't be ported to Linux and then asking it to install 3.13 would run `sudo emerge --ask dev-lang/python:3.13`).

My suspicion is that a lot of the people who protest loudly on behalf of "Linux distros" aren't really *involved* in Linux distros and aren't necessarily fully aware of what Linux distros think.

(I will confess I'm happy you didn't put the `install` commands on `python` itself. For me this is a command-line parsing purity thing, because on Unix, it is legal and not unreasonable to run `python install` and expect it to run a script in the current directory. I see no problem with making `python script.py` run slowly due to first downloading and installing python, since that is is *behaviorally compatible* with running an actual classic cpython binary.)

is possible that the PyManager MSIX package may not be installable on all
continuous integration systems. No differences are anticipated between a package
installed by PyManager compared to one installed by Nuget.

Users using the embeddable distro may have to change to a new method for
discovering the URL to the packages, though the recommendation would be to use
PyManager to discover and install. No differences are anticipated due to the
change of installer, and the embeddable distro package would be identical to
today.

Security Implications
=====================

In this section we compare the security implications of the installer itself to
the existing installers. The implications of Python being installed on a machine
are out of scope, and the ability of a malicious user to execute the installer
is also out of scope.

The typical risk introduced by an installer is that an elevated install may make
changes to a system that allow a low-privileged user to later affect a
high-privilege user, for example, by inappropriately setting access control on
shared folders. PyManager only operates as the same privilege level as the user,

> and therefore cannot introduce any escalation path.

Reviewer 3: The answer to this is probably obvious, but I'm unfamiliar with the ecosystem so I'll say it anyway...

My presumption was that an even more typical risk is that (as an installer that fetches the content to be installed from the internet) your internet feed could be corrupted, the server hacked, or your connection hijacked to serve malicious installers. Of course, the existing website links on https://www.python.org/downloads/windows/ are vulnerable to the same thing, but Microsoft Authenticode signatures protect the installer.

After this PEP,
> The default index file is hosted on python.org, and contains install information including package URLs and hashes for all installable versions. An alternate index may be specified by the user or their administrator (see Configuration below).

Which tells me nothing about the security implications of using such an index. Are the installable versions code-signed as well? Is the index file code-signed? The hashes are irrelevant from a security perspective -- anyone could upload malicious hashes alongside a malicious archive -- unless the hashes themselves are first validated via code-signing.

TLS does provide some (limited) defense here.

This section should explain in more detail whether or not this is all safe, and what guarantees are available.

Author: Essentially the same security applies here as for PyPI, plus of course we code sign all our binaries already (and nothing in the runtimes is changing, so that remains in place).

I've made a note to make this clearer in the next revision.
